src/components/join-flow/join-flow.jsx

```diff
@@ -21,48 +25,151 @@ class JoinFlow extends React.Component {
+            }, (err, body, res) => {
+                this.setState({waiting: false}, () => {
+                    let errStr = '';
+                    if (!err && res.statusCode === 200) {
+                        if (body && body[0]) {
+                            if (body[0].success) {
+                                this.props.refreshSession();
+                                this.setState({
+                                    step: this.state.step + 1
+                                });
+                                return;
+                            }
+                            if (body[0].errors) {
+                                // body can include zero or more error objects, each
+                                // with its own key and description. Here we assemble
+                                // all of them into a single string, errStr.
+                                const errorKeys = Object.keys(body[0].errors);
+                                errorKeys.forEach(key => {
+                                    const val = body[0].errors[key];
+                                    if (val && val[0]) {
+                                        if (errStr.length) errStr += '; ';
+                                        errStr += `${key}: ${val[0]}`;
+                                    }
+                                });
+                            }
+                            if (!errStr.length && body[0].msg) errStr = body[0].msg;
+                        }
+                    }
+                    this.setState({
+                        registrationError: errStr ||
+                            `${this.props.intl.formatMessage({
+                                id: 'registration.generalError'
+                            })} (${res.statusCode})`
+                    });
+                });
+            });
+        });
+    }
+    handleAdvanceStep (newFormData) {
+        newFormData = newFormData || {};
         this.setState({
-            step: this.state.step + 1,
-            formData: defaults({}, formData, this.state.formData)
+            formData: defaults({}, newFormData, this.state.formData),
+            step: this.state.step + 1
         });
     }
     render () {
         return (
             <React.Fragment>
-                <Progression step={this.state.step}>
-                    <UsernameStep onNextStep={this.handleAdvanceStep} />
-                    <BirthDateStep onNextStep={this.handleAdvanceStep} />
-                    <GenderStep onNextStep={this.handleAdvanceStep} />
-                    <CountryStep onNextStep={this.handleAdvanceStep} />
-                    <EmailStep onNextStep={this.handleAdvanceStep} />
-                    <WelcomeStep
-                        email={this.state.formData.email}
-                        username={this.state.formData.username}
-                        onNextStep={this.handleAdvanceStep}
+                {this.state.registrationError ? (
```

Comment:
Write a test to confirm that if state.registrationerror exists, we show that and don't show progression. To do this, we'll need to shallow mount JoinFlow and call wrapper.setState().